cpp/src/strings/convert/convert_lists.cu

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/convert/convert_lists.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/string_view.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+namespace {
+
+// position of the element separator string (e.g. comma ',') within the separators column
+constexpr size_type separator_index = 0;
+// position of the enclosure strings (e.g. []) within the separators column
+constexpr size_type left_brace_index  = 1;
+constexpr size_type right_brace_index = 2;
+
+/**
+ * @brief Pending separator type for `stack_item`
+ */
+enum class item_separator { NONE, ELEMENT, LIST };
+
+/**
+ * @brief Stack item used to manage nested lists.
+ *
+ * Each item includes the current range and the pending separator.
+ */
+struct alignas(8) stack_item {
```

Comment:
Since alignment is important here, should we specify the underlying type of the above enum? AFAICT it won't ever matter in practice since we'll get an `int` by default, just wanted to raise the question in case there's an edge case I'm not considering or if it helps the compiler optimize in any way by knowing that the underlying type is fixed.